### ISSUE TITLE
[Feature]: SM100 sparse indexer: stable log-probabilities and unclipped KL backward

### ISSUE THREAD

**Issue by C-TC** [NVIDIA/cudnn-frontend#1136]
### Before submitting

- [x] I searched existing issues and did not find a duplicate.

### Problem or use case

Could the SM100 DSA sparse indexer support standard, **unclipped distillation KL**, including stable log-probability output? We would like to agree on the intended API and semantics before proposing an implementation.

For a fixed, normalized teacher distribution `p`, student logits `z`, and `q = softmax(z)` over the same valid selected keys:

```text
KL(p || q) = sum_i p_i * (log(p_i) - log(q_i))
dKL/dz_i   = q_i - p_i
```

There are two related numerical limitations when using the existing sparse path for this objective:

1. **KL value:** sparse student score recomputation returns FP32 probabilities only. Once `q_i` underflows, reconstructing `log(q_i)` cannot recover the original finite log-probability. Clipping the reconstructed log also changes the loss.
2. **Gradient:** `indexer_backward_sm100.py` uses `CLIP_LOG_MIN = -100` and a `log_clip_mask` that suppresses the teacher contribution when the student probability falls below the threshold. For unclipped KL, as `q_i -> 0`, the gradient should approach **`-p_i`**, not zero.

The current implementation and reference appear consistent with an intentionally **clipped-log objective**. Returning FP32 probabilities is also a valid softmax output contract. This request concerns supporting a different objective reliably, rather than silently changing the existing contract.

### Minimal mathematical example

Student logits `[-150, 0]` and teacher probabilities `[1, 0]` give an expected KL of approximately **150**, with logit gradient **`[-1, 1]`**. The first student probability rounds to zero in FP32, and the current clipped preprocessing instead gives approximately zero score gradient. Teacher entropy uses the usual `0 * log(0) = 0` limit.

The clipping can matter before complete probability underflow: `exp(-101)` is representable as an FP32 subnormal but is already below the `exp(-100)` cutoff.

This is an algebra/FP32 illustration, not a claimed supported two-element native kernel launch shape.

### Source and validation evidence

Source checked at upstream `develop` commit `ebe3bba998b8e5d392f13a6ec491f3dbe8b2d591`:

- [SM100 sparse score recomputation](https://github.com/NVIDIA/cudnn-frontend/blob/ebe3bba998b8e5d392f13a6ec491f3dbe8b2d591/python/cudnn/deepseek_sparse_attention/score_recompute/_interface_sm100.py#L299): probability output only.
- [SM100 score-gradient preprocessing](https://github.com/NVIDIA/cudnn-frontend/blob/ebe3bba998b8e5d392f13a6ec491f3dbe8b2d591/python/cudnn/deepseek_sparse_attention/indexer_backward/indexer_backward_sm100.py#L3644): `g = -max(p, eps) * I(q >= eps)`, then `dz = g - q * sum(g)`, with `eps = exp(-100)` (omitting the upstream scale).
- [Backward reference](https://github.com/NVIDIA/cudnn-frontend/blob/ebe3bba998b8e5d392f13a6ec491f3dbe8b2d591/test/python/fe_api/dsa/dsa_reference.py#L727): explicitly matches that clipped-log contract. The earlier #297 also describes clipped-log behavior.
- [SM90 low-level sparse score interface](https://github.com/NVIDIA/cudnn-frontend/blob/ebe3bba998b8e5d392f13a6ec491f3dbe8b2d591/python/cudnn/deepseek_sparse_attention/score_recompute/_interface_sm90.py#L233): already offers `output_log_probs`; this is a potential API precedent, not a claim of tested SM90 backward equivalence.

We previously reproduced the gradient mismatch on **GB200**, using BF16 operands, cuDNN frontend 1.29.0, and score/backward sources corresponding to upstream `6a2df0db92ad53582e3f3a0dbf07bb22b1ada6b1`. A synthetic test with query/key lengths 256, top-k 128, and indexer H32/D128 compared against an independent FP64 reference. Increasing student-logit spread via gate multipliers of 32/128 produced gate-gradient relative L2 errors of approximately **7.55% / 90.37%** against exact KL, versus 0.17% at multiplier 1.

A local prototype computes **`(q - p) * upstream * normalization`** directly, optionally returns log-probabilities from logits/LSE, and retains the existing backward GEMMs. It passed FP64-reference regressions for large logit spreads and signed upstream gradients, together with graph-replay checks.

Those are **historical GPU results**. The latest commit above was source-audited; neither the reproducer nor the prototype has been rerun on that commit for this issue. The mathematical example was checked separately on CPU.

### Proposed solution / questions

Would an explicit unclipped-KL mode be appropriate, or would you prefer exposing the primitives needed for callers to compose it?

- For the loss, optional stable `log_q = z - logsumexp(z)` output, or logits/LSE from which to compute it.
- For the gradient, a direct `q - p` path with upstream scaling and normalization, or a supported interface to supply that score gradient to the backward GEMMs. This avoids both `1/q` and reconstructing log-probabilities from probabilities.
- Preserve existing clipped behavior for callers that intentionally use it.

Either approach should be tested against an independent FP64 oracle for ordinary and large logit spreads, valid-length masking, and positive/negative/zero upstream gradients. The desired outcome is accurate KL values and gradients throughout the supported numerical range while retaining the existing GEMM performance.
